accepted/2024/repo-level-lock-file.md

```diff
@@ -0,0 +1,84 @@
+﻿# ***Repository Level Lock File***
+
+- Author: [@CEbbinghaus](https://github.com/CEbbinghaus) 
+- Issue: [Nuget/Home#12409](https://github.com/NuGet/Home/issues/12409) 
+
+## Summary
+
+Applications consisting of multiple solutions (using assembly references between projects) cannot control their transitive dependencies by using Nuget. The current lock file is on a per-project basis rather than an application/repository basis. Given a `Directory.packages.props` file, each project could resolve different versions of transitive dependencies, which a single central lock file would solve by allowing consistency between package versions across multiple solutions.
+
+
+## Motivation 
+
+Monorepositories with multiple solutions utilizing assembly references that deploy all their code together as a single application are unable to use Nuget (with or without CPM) due to transitive version conflicts. This is due to NuGet being able to resolve different versions of transitive dependencies for each individual build, as they occur in isolation from one another. A single central lock file allows all transitive dependencies to have locked versions defined at the root, thereby eliminating third-party dependency management software such as Paket.
+
+
+## Explanation
+
+### Functional explanation
+
+<!-- Explain the proposal as if it were already implemented, and you're teaching it to another person. -->
+<!-- Introduce new concepts, functional designs with real-life examples, and low-fidelity mockups or pseudocode to show how this proposal would look. -->
+
+The `Directory.packages.props` now uses the`<NuGetLockFilePath>[path]</NuGetLockFilePath>` property within a `<PropertyGroup>` to specify its lock file path. It becomes the source of truth for all builds. Project level lock files are no longer required unless a package version is overridden.
+
+```xml
+<Project>
+    <PropertyGroup>
+        <ManagePackageVersionsCentrally>true</ManagePackageVersionsCentrally>
+        <NuGetLockFilePath>lockfile.json</NuGetLockFilePath>
+    </PropertyGroup>
+    <ItemGroup>
+        <PackageVersion Include="Stuff" Version="1.0.0" />
+    </ItemGroup>
+</Project>
+```
+
+This lock file is generated during any restore operation during the evaluation of the `Directory.packages.props` file. This allows the central lock file to be generated by any project as all of the dependencies are isolated.
```

Comment:
The biggest challenges delivering this feature are the preexisting msbuild paradigms and we're not really tackling that here. 

We need to dive deeper into the technical complexities to understand the feasibility of simply having a lock file.

How would the scope of the msbuild based restore operation be defined? Currently restore is either project based or solution based. Projects don't know about all other projects. 

Here's a practical example: 

You have 2 projects and 3 packages.

Project 1:
```xml
<PackageReference Include="A" />
```

Project 2:
```
<PackageReference Include="B" />
```

DPP
```xml
<PackageVersion Include="A" Version="1.0.0" />
<PackageVersion Include="B" Version="1.0.0" />
```

A 1.0.0 -> C 1.0.0
B 1.0.0 -> C 2.0.0

Without pinning:

- Project 1: A 1.0.0 and C 1.0.0
- Project 2: B 1.0.0 and C 2.0.0

These projects aren't necessarily correlated so the only way for them to know that they're conflicting with C is to either: 
- Consider what each project references each restore
- Enforce version declaration for every package version encountered. 

Today, a project only knows about itself and it's closure, so this would be a departure in how restore/build work, one that'd come with some perf considerations. 

The 2nd idea could theoretically be solved with transitive pinning. 

You can declare `C 2.0.0` in DPP, enable transitive pinning and then the versions would be resolved would be equivalent. 
You'd basically need to exhaustively declare *all* versions.

Within the current toolset, The central lock file as proposed is really a more algorithmically advanced transitive pinning that gets applied across *all* projects. 

Another challenge is package creation for any projects that have a central lock file. 

NuGet works really hard to make projects and the package versions of them behave as closely as possible. 
What this means is that every PackageReference you have becomes a dependency in a generated package.
To avoid runtime issues, transitive pinning elevates those transitively pinned packages to top level dependencies. The practical example is Project 1 from above, where it could unwittingly take a dependency on C 2.0.0, and then if it gets published as a package with only its direct dependencies lead to runtime failures later. 
I understand that's not really a relevant scenario here, but it is a concept that the design needs to account for.

Reply:
> the central lock file as proposed is really a more algorithmically advanced transitive pinning that gets applied across all projects

Isn't that exactly what a lock file does, by specifying the entire dependency graph and the versions used of each dependency?

> The practical example is Project 1 from above, where it could unwittingly take a dependency on C 2.0.0, and then if it gets published as a package with only its direct dependencies lead to runtime failures later.

I assume we would need to keep track of which dependencies are direct and which are transitive, and `dotnet pack` would only look at direct dependencies when building a package. A transitive dependency can always be elevated to a direct dependency.

Reply:
> Isn't that exactly what a lock file does, by specifying the entire dependency graph and the versions used of each dependency?

There's a key distinction. All version resolution is project based. 
Lock files don't change the resolution. They skip it. 

The idea for central lock files makes the version resolution more involved and requires to account for transitive packages declared from all projects, not just the current one. 

You're not just asking for a file, you're asking for a completely new resolver. Beyond scoping, the resolver part is the most challenging technical part. 

> I assume we would need to keep track of which dependencies are direct and which are transitive, and dotnet pack would only look at direct dependencies when building a package. A transitive dependency can always be elevated to a direct dependency.

"dotnet pack would only look at direct dependencies when building a package" -> This doesn't work in all scenarios. 
Here's an example: 

Project 1 -> Package A 1.0.0 -> Package B 1.0.0
Regular resolution leads to A 1.0.0, B 1.0.0. 
When the project is packaged, it only needs A 1.0.0 as a dependency, since at minimum B 1.0.0 will be resolved due to A's dependency. 

Central lock file example could bump package B to 2.0.0. 
Project 1 code ends up coding against B 2.0.0. 
If the Project 1 package ends up declaring A 1.0.0, then when anyone installs that package generated from project 1, they'd get B 1.0.0, because no one knows about 2.0.0 in the Project 1 graph, it's there due to how other projects work. If we don't elevate B 2.0.0 to a direct dependency, then the consumer of Project 1's generated package would end up with runtime failures.

Reply:
Some general food for thought for the concept of centralized lock files. 

In general, you probably only need to consolidate the versions being used across your runnable projects. Your services, your tools etc, not across your libraries as well (and never downgrade ofc).
Just because you make your projects reference a specific version, that doesn't mean they'll all work. You could be referencing dependencies that reference different versions of a particular package, so consolidation is naturally going to happen in that case. 

There's of course the theory for handling breaking changes in dependencies, but that'd only work if you control all libraries that reference a package with a breaking change, in which case regular CPM would work just fine.

Reply:
I prefer the idea of promoting any transitive dependency at a version outside the already resolved range into a primary dependency so that any downstream library consumers don't need to think about it